pycape/llms/llms.py

```diff
@@ -0,0 +1,243 @@
+        """Create or load a :class:`~token.Token`.
+
+        Args:
+            token: Filepath to a token file, or the raw token string itself.
+
+        Returns:
+            A :class:`~token.Token` that can be used to access users' deployed Cape
+            functions.
+
+        Raises:
+            TypeError: if the ``token`` argument type is unrecognized.
+        """
+        token_out = None
+        if isinstance(token, pathlib.Path):
+            tokenfile = token
+            return tkn.Token.from_disk(tokenfile)
+
+        if isinstance(token, str):
+            # str could be a filename
+            if len(token) <= 255:
+                token_as_path = pathlib.Path(token)
+                token_out = _try_load_token_file(token_as_path)
+            return token_out or tkn.Token(token)
+
+        if isinstance(token, tkn.Token):
+            return token
+
+        raise TypeError(f"Expected token to be PathLike or str, found {type(token)}")
+
+    async def chat_completions(
+        self,
+        messages: Union[str, List[Dict[str, Any]]],
+        token: str,
+        stream=True,
+        model="llama",
+        pcrs=None,
+    ):
+        await self._connect("/v1/ws/chat/completions", token, pcrs=pcrs)
+
+        aes_key = os.urandom(32)
+        user_key = base64.b64encode(aes_key).decode()
+
+        data = crypto.envelope_encrypt(
+            self.ctx.public_key,
+            {"request": {"messages": messages, "stream": stream}, "user_key": user_key},
+        )
+        data = base64.b64encode(data).decode()
+
+        msg = WSMessage(
+            msg_type=WSMessageType.CHAT_COMPLETIONS_REQUEST,
+            data=data,
+        ).model_dump_json()
+
+        await self.ctx.websocket.send(msg)
+
+        async for msg in self.ctx.websocket:
+            msg = WSMessage.model_validate_json(msg)
+            if msg.msg_type != WSMessageType.STREAM_CHUNK:
+                raise Exception(
+                    f"expected {WSMessageType.STREAM_CHUNK} not {msg.msg_type}"
+                )
+
+            dec = crypto.aes_decrypt(
+                base64.b64decode(msg.data["data"].encode()), aes_key
+            )
+
+            yield dec.decode()
+
+    @property
+    def ctx(self):
+        if self._ctx is None:
+            raise Exception("must call _connect first")
+
+        return self._ctx
+
+    async def _connect(self, endpoint, token, pcrs=None):
+        endpoint = self._url + endpoint
+        self._root_cert = self._root_cert or attest.download_root_cert()
+        self._ctx = _Context(
+            endpoint=endpoint,
+            auth_token=token.raw,
+            root_cert=self._root_cert,
+        )
+        attestation_doc = await self._ctx.bootstrap(pcrs)
+
+        return attestation_doc
+
+
+class WSMessageType(str, Enum):
+    NONCE = "nonce"
+    ATTESTATION = "attestation"
+    STREAM_CHUNK = "stream_chunk"
+    CHAT_COMPLETIONS_REQUEST = "chat_completion_request"
```

Comment:
How would we version this protocol? I can see the protocol changing like our old ones did.

Reply:
We can use the subprotocol parameters when connecting to the server https://stackoverflow.com/questions/67436517/what-is-a-websocket-subprotocol

Reply:
Which ends up just setting a header.

Reply:
Yeah, we can add that later when the protocol changes